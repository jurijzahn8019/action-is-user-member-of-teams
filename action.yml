name: "Is User a Member of any Team"
description: |
  This action will take the username and a list of team names in format <org>/<team>
  And then perform an api Query to fetch the team metadata and evaluate whether the user
  is a member of any of them.

  Then it will set a boolean result as output
inputs:
  token:
    description: "Github Api token with repo scope, e.g. secrets.GITHUB_TOKEN"
    required: true
  username:
    description: |
      Username of the github user to check: anyuser3445
      Can also be comma separated list of users user1,user2

      This action returns true if any user is a member of any team
      First match results to true
    required: true
  teams:
    description: "Comma-sparated list of teams to check membership: org/team1,org/team2"
    required: true
  varName:
    description: "If given, then action will export the result as this env Var"
    required: false
    default: USER_IS_MEMBER
outputs:
  ismember:
    description: "true if user is member of one or more teams"
  userName:
    description: Username of the first found user who is member of a team 
  teamName:
    description: Team name in case the user's team was found
runs:
  using: node16
  main: dist/index.js
