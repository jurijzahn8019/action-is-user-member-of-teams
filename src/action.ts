/* eslint-disable camelcase */
import {
  getInput,
  setFailed,
  setOutput,
  exportVariable,
  info,
} from "@actions/core";
import { getOctokit } from "@actions/github";
import debug from "debug";

const dbg = debug("action-is-user-member-of-teams:index");

export type GqlTeamMemberResult = {
  name: string;
};

export type GqlTeamResult = {
  name: string;
  members: { nodes: GqlTeamMemberResult[] };
};

export type GqlOrgResult = Record<string, { nodes: GqlTeamResult[] }>;

export type GqlResult = Record<string, GqlOrgResult>;

export async function run(): Promise<void> {
  dbg("Check whether user is memebr of teams");
  try {
    dbg("Retrieve inputs");
    const token = getInput("token", { required: true });
    const usernames = getInput("username", { required: true })
      .split(",")
      .map((u) => u.trim());
    const varName =
      getInput("varName", { required: false }) || "USER_IS_MEMBER";
    const teams = getInput("teams", { required: true })
      .split(",")
      .map((t) => {
        const [org, team] = t.trim().split("/");
        return { org, team };
      });
    dbg("Check users: %o in teams: %o", usernames, teams);

    dbg("Fetch teams by name and org");
    const byOrg = [] as { name: string; teams: string[] }[];
    teams.forEach((t) => {
      let org = byOrg.find((o) => o.name === t.org);
      if (!org) {
        org = { name: t.org, teams: [] };
        byOrg.push(org);
      }

      if (!org.teams.find((n) => n === t.team)) {
        org.teams.push(t.team);
      }
    });

    const oqs = byOrg.map(({ name, teams: orgTeams }) => {
      const orgKey = name.replace(/\W/g, "_");

      const tqs = orgTeams.map((t) => {
        const teamkey = t.replace(/\W/g, "_");
        return `${teamkey}: teams(first: 100, query: "${t}") { ...Teams }`;
      });

      return `${orgKey}: organization(login: "${name}") { name ${tqs.join(
        "\n"
      )} }`;
    });

    const query = `
    fragment Teams on TeamConnection {
      nodes { name members { nodes { name: login } } }
    }
    query { 
      ${oqs.join("\n")} 
    }`;

    const client = getOctokit(token);

    const data = await client.graphql<GqlResult>(query);

    dbg("Process result data: %o", data);
    const users = Object.values(data)
      .map((org) => {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const { name, ...tdata } = org;
        return Object.values(tdata).map((orgTeams) =>
          orgTeams.nodes.map((t) =>
            t.members.nodes.map((user) => ({
              org: org.name,
              team: t.name,
              user: user.name,
            }))
          )
        );
      })
      .flat(4);

    dbg("Apply team filter to the users list");
    const filtered = users.filter(({ team }) =>
      teams.some((t) => team === t.team)
    );

    const res = filtered.find((u) => usernames.find((un) => un === u.user));
    if (res) {
      dbg("User %s is member of given team: %s", res.user, res.team);
      info(`Found User ${res.user} is member of team ${res.team}`);

      setOutput("orgName", res.org);
      setOutput("teamName", res.team);
      setOutput("userName", res.user);
    } else {
      dbg("Users %s not member of any of given teams", usernames.join(","));
      info(
        `Users ${usernames.join(",")} not member of ${teams
          .map((t) => `${t.org}/${t.team}`)
          .join(" or ")}`
      );

      setOutput("orgName", undefined);
      setOutput("teamName", undefined);
      setOutput("userName", undefined);
    }

    setOutput("ismember", !!res);
    exportVariable(varName, !!res);
  } catch (e) {
    dbg("Failed:", e);
    setFailed((e as Error).message);
  }
}
