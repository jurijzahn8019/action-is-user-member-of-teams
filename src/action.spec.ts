/* eslint-disable @typescript-eslint/no-explicit-any */
import { getInput, info, setFailed, setOutput } from "@actions/core";
import { getOctokit } from "@actions/github";
import { GqlOrgResult, GqlResult, run } from "./action";

jest.mock("@actions/core");
jest.mock("@actions/github");

const getInputMock = getInput as jest.MockedFunction<typeof getInput>;
const setFailedMock = setFailed as jest.MockedFunction<typeof setFailed>;
const getOctokitMock = getOctokit as jest.MockedFunction<typeof getOctokit>;
const infoMock = info as jest.MockedFunction<typeof info>;
const setOutputMock = setOutput as jest.MockedFunction<typeof setOutput>;

const octokit = {
  graphql: jest.fn() as jest.Mock<Promise<GqlResult>, any>,
};

describe("action", () => {
  beforeEach(() => {
    getOctokitMock.mockReturnValue(octokit as any);
    getInputMock.mockReturnValueOnce("THE TOKEN");
    getInputMock.mockReturnValueOnce("user1,user2");
    getInputMock.mockReturnValueOnce("");
    getInputMock.mockReturnValueOnce("foo/bar,foo/baz,baz/bar,baz/bar");
    octokit.graphql.mockResolvedValue({
      non: { bar: { nodes: [] } },
      foo: {
        bar: {
          nodes: [
            {
              name: "bar",
              members: { nodes: [{ name: "user4" }, { name: "user1" }] },
            },
          ],
        },
      },
    });
  });

  it("Happy Path", async () => {
    await expect(run()).resolves.toBeUndefined();

    expect(getInputMock).toHaveBeenCalledTimes(4);
    expect(getInputMock.mock.calls.map((c) => c[0])).toEqual([
      "token",
      "username",
      "varName",
      "teams",
    ]);

    expect(getOctokitMock).toHaveBeenCalledTimes(1);
    expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN");

    expect(octokit.graphql.mock.calls[0][0]).toMatchSnapshot("Query");

    expect(setOutputMock).toHaveBeenCalledTimes(4);
    expect(setOutputMock.mock.calls.map((c) => c.join(": "))).toMatchSnapshot(
      "Output"
    );

    expect(infoMock).toHaveBeenCalledTimes(1);
    expect(infoMock).toHaveBeenCalledWith(
      "Found User user1 is member of team bar"
    );

    expect(setFailedMock).not.toHaveBeenCalled();
  });

  it("Should fail on error", async () => {
    octokit.graphql.mockRejectedValue(new Error("The FOO"));

    await expect(run()).resolves.toBeUndefined();

    expect(getInputMock).toHaveBeenCalledTimes(4);
    expect(getInputMock.mock.calls.map((c) => c[0])).toEqual([
      "token",
      "username",
      "varName",
      "teams",
    ]);

    expect(getOctokitMock).toHaveBeenCalledTimes(1);
    expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN");

    expect(octokit.graphql).toHaveBeenCalledTimes(1);

    expect(setOutputMock).not.toHaveBeenCalled();
    expect(infoMock).not.toHaveBeenCalled();

    expect(setFailedMock).toHaveBeenCalledTimes(1);
    expect(setFailedMock).toHaveBeenCalledWith("The FOO");
  });

  it("Should report negative", async () => {
    octokit.graphql.mockResolvedValue({
      non: { bar: { nodes: [] } },
      foo: {
        bar: {
          nodes: [
            {
              name: "non",
              members: { nodes: [{ name: "user666" }] },
            },
          ],
        },
      },
    });

    await expect(run()).resolves.toBeUndefined();

    expect(getInputMock).toHaveBeenCalledTimes(4);
    expect(getInputMock.mock.calls.map((c) => c[0])).toEqual([
      "token",
      "username",
      "varName",
      "teams",
    ]);

    expect(getOctokitMock).toHaveBeenCalledTimes(1);
    expect(getOctokitMock).toHaveBeenCalledWith("THE TOKEN");

    expect(octokit.graphql.mock.calls[0][0]).toMatchSnapshot("Query");

    expect(setOutputMock).toHaveBeenCalledTimes(4);
    expect(setOutputMock.mock.calls.map((c) => c.join(": "))).toMatchSnapshot(
      "Output"
    );

    expect(infoMock).toHaveBeenCalledTimes(1);
    expect(infoMock).toHaveBeenCalledWith(
      "Users user1,user2 not member of foo/bar or foo/baz or baz/bar or baz/bar"
    );

    expect(setFailedMock).not.toHaveBeenCalled();
  });
});
