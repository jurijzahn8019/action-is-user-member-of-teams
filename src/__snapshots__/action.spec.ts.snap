// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action Happy Path: Output 1`] = `
Array [
  "orgName: ",
  "teamName: bar",
  "userName: user1",
  "ismember: true",
]
`;

exports[`action Happy Path: Query 1`] = `
"
    fragment Teams on TeamConnection {
      nodes { name members { nodes { name: login } } }
    }
    query { 
      foo: organization(login: \\"foo\\") { name bar: teams(first: 100, query: \\"bar\\") { ...Teams }
baz: teams(first: 100, query: \\"baz\\") { ...Teams } }
baz: organization(login: \\"baz\\") { name bar: teams(first: 100, query: \\"bar\\") { ...Teams } } 
    }"
`;

exports[`action Should report negative: Output 1`] = `
Array [
  "orgName: ",
  "teamName: ",
  "userName: ",
  "ismember: false",
]
`;

exports[`action Should report negative: Query 1`] = `
"
    fragment Teams on TeamConnection {
      nodes { name members { nodes { name: login } } }
    }
    query { 
      foo: organization(login: \\"foo\\") { name bar: teams(first: 100, query: \\"bar\\") { ...Teams }
baz: teams(first: 100, query: \\"baz\\") { ...Teams } }
baz: organization(login: \\"baz\\") { name bar: teams(first: 100, query: \\"bar\\") { ...Teams } } 
    }"
`;
